fireant/database/base.py

```diff
@@ -1,40 +1,100 @@
+from functools import wraps
+
 import pandas as pd
+import time
+
 from pypika import (
     Query,
     enums,
     functions as fn,
     terms,
 )
+from .logger import (
+    query_logger,
+    slow_query_logger,
+)
+
+
+def log(func):
+    @wraps(func)
+    def wrapper(database, query):
+        start_time = time.time()
+        query_logger.debug(query)
+
+        result = func(database, query)
+
+        duration = round(time.time() - start_time, 4)
+        query_log_msg = '[{duration} seconds]: {query}'.format(duration=duration,
+                                                               query=query)
+        query_logger.info(query_log_msg)
+
+        if duration >= database.slow_query_log_min_seconds:
+            slow_query_logger.warning(query_log_msg)
+
+        return result
+
+    return wrapper
+
+
+@log
+def fetch(database, query):
+    with database.connect() as connection:
+        cursor = connection.cursor()
+        cursor.execute(query)
+        return cursor.fetchall()
+
+
+@log
+def fetch_data(database, query):
```

Comment:
(Commenting now as I was away when it was merged)

I noticed this function called in each thread executing a query, which creates a new database connection per query. I think that we will quickly hit any database connection limits unless we share the connection object for all of the threads executing the query.

Reply:
will raise an issue